### PR TITLE
Feature: "critical" resource meta-attribute and "influence" colocation constraint option

### PR DESCRIPTION
**Pull request by kgaillot**
This allows "noncritical" resources that do not cause the resource they depend on to move if they fail.

### CODE REVIEW

**Comment by kgaillot** [comment]
@clumens , can you review this? The "Feature: scheduler" commit will be challenging because it is deep in the bowels of the scheduler, but the rest should be reasonably straightforward to follow.

File: lib/pacemaker/pcmk_sched_constraints.c [hunk 1]
```diff
@@ -2279,8 +2281,38 @@ unpack_rsc_order(xmlNode * xml_obj, pe_working_set_t * data_set)
...
+ *         or value of resource's critical option if string is NULL or invalid
+ */
+static bool
+unpack_influence(const char *coloc_id, const pe_resource_t *rsc,
+                 const char *influence_s)
+{
+    if (influence_s != NULL) {
+        int influence_i = 0;
+
+        if (crm_str_to_boolean(influence_s, &influence_i) < 0) {
```
Comment (clumens) [inline lib/pacemaker/pcmk_sched_constraints.c:2306]: I might be missing some subtle distinction here, but couldn't you use `if (!crm_is_true(influence_s))` and skip the intermediate variable?

File: lib/pacemaker/pcmk_sched_utils.c [hunk 2]
```diff
@@ -765,34 +765,3 @@ pcmk__create_history_xml(xmlNode *parent, lrmd_event_data_t *op,
...
- * \internal
- * \brief Check whether a colocation constraint should apply
- *
- * \param[in] rsc            Resource of interest (for logging)
- * \param[in] colocation     Colocation constraint to check
- * \param[in] promoted_only  If true, constraint applies if right-hand is promoted
- */
-bool
-pcmk__colocation_applies(pe_resource_t *rsc, pcmk__colocation_t *colocation,
-                         bool promoted_only)
```
Comment (clumens) [inline lib/pacemaker/pcmk_sched_utils.c:779]: Alas, pcmk__colocation_applies, we hardly knew ye.

File: include/pcmki/pcmki_scheduler.h [hunk 3]
```diff
@@ -107,4 +107,30 @@ extern gboolean show_scores;
...
+ *                        this can be a particular instance)
+ *
+ * \return true if colocation influence should be effective, otherwise false
+ */
+static inline bool
+pcmk__colocation_has_influence(const pcmk__colocation_t *colocation,
+                               const pe_resource_t *rsc)
+{
+    if (rsc == NULL) {
+        rsc = colocation->rsc_rh;
```
Comment (clumens) [inline include/pcmki/pcmki_scheduler.h:127]: There's no way that colocation->rsc_rh could be NULL, is there?

File: lib/pacemaker/pcmk_sched_constraints.c (hunk 1 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_constraints.c:2306]: Necessary to use pe_rsc_critical as the default

File: lib/pacemaker/pcmk_sched_utils.c (hunk 2 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_utils.c:779]: lol

File: include/pcmki/pcmki_scheduler.h (hunk 3 above)
Comment (kgaillot) [reply include/pcmki/pcmki_scheduler.h:127]: No, colocations can only be created by pcmk__new_colocation(), which won't create one without it

**Pull request merged**